include/boost/windows_sspi/stream.hpp

```diff
@@ -68,61 +68,43 @@ template <typename NextLayer, typename ConstBufferSequence> struct async_write_i
     : m_next_layer(next_layer)
-    , m_buffer(buffer)
+    , m_buffers(buffers)
     , m_sspi_impl(std::move(sspi_impl)) {
   }
 
   template <typename Self> void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
     BOOST_ASIO_CORO_REENTER(*this) {
-      while(m_sspi_impl->decrypt.decrypted_data.empty()) {
-        // TODO: Find some way to make the sspi_impl, the decrypt
-        // function or something else be responsible for keeping track
-        // of state and buffer(s)
-        // TODO: Fix so overflow cannot happen (we don't need to read
-        // more data unless DecryptMessage asks us to).
-        BOOST_ASSERT(m_sspi_impl->decrypt.encrypted_data.size() < 0x10000);
-        m_message.resize(0x10000 - m_sspi_impl->decrypt.encrypted_data.size());
+      while(m_sspi_impl->decrypt() == detail::sspi_decrypt::state::data_needed) {
+        // TODO: Use a fixed size buffer instead
+        m_input.resize(0x10000);
         BOOST_ASIO_CORO_YIELD net::async_read(m_next_layer,
-                                              net::buffer(m_message),
+                                              net::buffer(m_input),
                                               std::move(self));
-        if (ec && length == 0 && m_sspi_impl->decrypt.decrypted_data.empty()) {
-          self.complete(ec, 0);
-          return;
-        }
-        m_message.resize(length);
-        auto sc = m_sspi_impl->decrypt(m_message);
-        m_message.clear();
-        if (sc == SEC_E_INCOMPLETE_MESSAGE) {
-          continue;
-        }
-        if (sc == SEC_I_CONTEXT_EXPIRED) {
-          // TODO: Shutdown the TLS context gracefully (implement
-          // async_shutdown), then return EOF.
-          self.complete(net::error::eof, 0);
-          return;
-        }
-        if (sc != SEC_E_OK) {
-          self.complete(error::make_error_code(sc), 0);
-          return;
-        }
+        m_sspi_impl->decrypt.put({m_input.begin(), m_input.begin() + length});
+        m_input.clear();
+        continue;
       }
 
-      std::size_t to_return = std::min(net::buffer_size(m_buffer), m_sspi_impl->decrypt.decrypted_data.size());
-      std::size_t bytes_copied = net::buffer_copy(m_buffer, net::buffer(m_sspi_impl->decrypt.decrypted_data, to_return));
-      boost::ignore_unused(bytes_copied);
-      BOOST_ASSERT(bytes_copied == to_return);
-      m_sspi_impl->decrypt.decrypted_data.erase(m_sspi_impl->decrypt.decrypted_data.begin(), m_sspi_impl->decrypt.decrypted_data.begin() + to_return);
-      self.complete(boost::system::error_code{}, to_return);
+      if (m_sspi_impl->decrypt() == detail::sspi_decrypt::state::error) {
+        ec = boost::error::make_error_code(m_sspi_impl->decrypt.error_code);
+        self.complete(ec, 0);
+        return;
+      }
+
+      const auto data = m_sspi_impl->decrypt.get(net::buffer_size(m_buffers));
+      std::size_t bytes_copied = net::buffer_copy(m_buffers, net::buffer(data));
```

Comment:
I can't help thinking that decrypting directly into user buffers will avoid a redundant copy.